Neutrinos ben

nebblu opened the pull request:
What I've done:
	•	Removed all warning errors from BSPT,BSPTN,SpecialFunctions,SPT,SCOL,HALO files. -Added in Bill's commits to SPT and SpecialFunctions (am re adding these because have removed all warning errors too!)
	•	Added in a flat to scol_init and scol_initp functions in HALO which indicate whether passed spectrum is LCDM at z=0 (false) or MGCAMB transfer at desired z (true). This affects the variance calculation.
	•	Added in new functions react_init_nu and react_init_nu that intialise k_star and mathcal{E} for the reaction computation in the presence of massive neutrinos.
	•	Added in function reaction_nu which calculates the reaction in the presence of massive neutrinos.
	•	Added in initialise function that initialises everything for the reaction with a single function call.
	•	Added a new class to LinearPS which takes as input the matter spectrum at desired z and outputs the spline needed for all internal computations.


IGNORE THE LAST 4 commits where I added and removed some validation files that we don't need right now.

nebblu committed:
Added a test example file to do output testing. Added a new class to linearPS which takes as input a CAMB linear spectrum. Added a folder with some MGCAMB outputs for validation

nebblu committed:
Updated HALO class to include all linear spectra. Added reaction functions for neutrinos + mg

nebblu committed:
Removed all warning errors from BSPT,BSPTN,SCOL,SpecialFunctions,SPT. Also including Bill's update to SPT and SpecialFunctions.

nebblu committed:
Added in 4 new functions to HALO class: react_init_nu, react_init_nu2, reaction_nu and initialise. The first two initialise k_star and mathcal E ffor the reaction. The 3rd is the reaction itself with neutrinos inlcuded. The fourth is a function that initialises everytthing needed for the reaction computation. I have also updated scol_init and scol_initp with a flag, mgcamb, that declares whethher the passed power spectra are at the desired redshift or at z=0 and need to be rescaled to the desired redshift.

nebblu committed:
Addedd example file that ttests the new implementations vs the old ones for the case of omega_nu =0

nebblu committed:
Added new class to LinearPS whhich takes as input the desired power spectrum and splines it for use in all other libraries

nebblu committed:
Test example for validating new functions

nebblu committed:
Delete test_transfer_z1.dat

nebblu committed:
removed some stuff

nebblu committed:
Delete pk_test.cpp

Alkistis commented:
Thanks Ben -- FYI I'm starting going through these now.

nebblu commented:
Hi Alkistis,

Thanks. It's a bit of a mess but won't have time to comment thoroughly
before I leave for Sweden today.

The only really relevant files are the src files:  HALO.cpp and HALO.h. In
all other commits I've basically removed warning messages (redundant
variables, undeclared variables, etc).

In SPT.cpp, SPT.h,  SpecialFunctions.cpp and SpecialFunctions.h I've also
included Bill's edits.

Cheers,
Ben

On Tue, Jul 14, 2020 at 8:02 AM Alkistis Pourtsidou <
notifications@github.com> wrote:

> Thanks Ben -- FYI I'm starting going through these now.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/nebblu/ReACT/pull/10#issuecomment-657986048>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AK52VTPYYGHEGJCU2OCOXHLR3PYH3ANCNFSM4OZFFNYQ>
> .
>

Alkistis reviewed (commented):
Comment on commit 0a8ba2a: these test files and code are removed in a later commit so nothing to act upon.

Alkistis commented on reactions/include/Copter/HALO.h:
```diff
@@ -5,12 +5,11 @@
 
 // limits on mass exponents
```
Suggestion: improve comment to explain how Mmin, Mmax are/should be chosen.

nebblu replied:
added comments on how and why these values are chosen

Alkistis commented on reactions/include/Copter/HALO.h:
```diff
@@ -5,12 +5,11 @@
 public:
-  HALO(const Cosmology& C, const PowerSpectrum& P_l, real epsrel = 1e-4);
-
+  HALO(const Cosmology& C, const PowerSpectrum& P_l, const PowerSpectrum& P_cb,const PowerSpectrum& P_nu, real epsrel = 1e-4);
```
Will this work if someone wants massless neutrinos?

nebblu replied:
Yes, in the massless case the user just needs to supply the linear spectrum for total matter in place of P_cb and P_nu. This should be added to the readme file. I've added a comment in HALO.h and HALO.cpp

Alkistis commented on reactions/src/HALO.cpp:
```diff
@@ -56,8 +56,8 @@ void react_error_halo(const char* msg) {
 /*Load the linear power spectrum */
-HALO::HALO(const Cosmology& C_, const PowerSpectrum& P_l_, real epsrel_)
-: C(C_), P_l(P_l_)
+HALO::HALO(const Cosmology& C_, const PowerSpectrum& P_l_, const PowerSpectrum& P_cb_, const PowerSpectrum& P_nu_, real epsrel_)
```
Same as above: will this work for massless nu (i.e., what if I want to recover the previous case).

nebblu replied:
added comment. See above.

Alkistis commented on reactions/src/HALO.cpp:
```diff
@@ -116,9 +132,35 @@ void HALO::scol_init(double vars[]) const{
+ // select what variance to use  (if input P(k) is LCDM at z=0, vars[6]=0, if P(k) if modified at z=z, vars[6]=1)
+ double sig1,sig2;
+
+  if (vars[6] < 0.5) {
```
is this a bit dangerous perhaps? why 0.5 and not a very small number since the "truth" is 0 or 1?

nebblu replied:
Agree. Made it a bool variable/flag (mgcamb) now so it's either true or false.

Alkistis commented on reactions/src/HALO.cpp:
```diff
@@ -149,8 +191,12 @@ for(int i = 0; i< loop_N; i++){
    scol.myscol(myscolparams, vars[0], vars[1], Rth, sig1, sig2, pars, 2, yenvf);
 
 // calculate sigma^2
+  if (vars[6]<0.5) {
```
same comment as before, I see this choice is made in several places so if changed it has to be changed everywhere it appears

nebblu replied:
see above.

Alkistis commented on reactions/src/HALO.cpp:
```diff
@@ -671,6 +715,147 @@ double HALO::plinear_cosmosis(double k) const {
+  double pspt, psptp, p1h, p1hp, plm, plcb, plnu, psptcbv, argument;
+
+  // fv
+  double fv  = vars[1]/vars[7];
```
perhaps an error message here if this ends up being 0 or bigger than 1 or infinity due to a user error? does it give the correct massless case results if it is 0?

Alkistis replied:
Again this appears in various places so if we put a flag/error we should do it everywhere.

nebblu replied:
So this was an error. It should be vars[7]/vars[1]  where vars[1] is the cb density fraction which should never be 0. vars[7] can safely be 0 (making fv =0 and reducing the reaction to the massless neutrino case). 

I think error flags for cosmological parameter consistency should be at the top level since they enter a host of functions later on. What do you think?

Alkistis replied:
yeah makes sense to be at the top level

Alkistis commented on reactions/src/HALO.cpp:
```diff
@@ -671,6 +715,147 @@ double HALO::plinear_cosmosis(double k) const {
+  double fvt2  = pow(fvt,2);
+
+/// mathcal E///
+  bigE = pow2(1.-fv)*one_halo(0.01,vars)/one_halop(0.01,vars);
```
These 0.01 and 0.06 k values choices, wouldn't it be better if they were named variables? Looking at the latest ReACT paper I'd name these smt like k_lim, k_bar.

nebblu replied:
Agree. Have added them as variables with a comment and reference.

Alkistis commented on reactions/src/HALO.cpp:
```diff
@@ -671,6 +715,147 @@ double HALO::plinear_cosmosis(double k) const {
+  psptp = spt.PLOOPn2(8, vars, 0.06, 1e-3)+ p1hp ;
+
+// cbv terms
+  psptcbv = pspt*plnu;
```
I'd expect this to be `psptcbv = sqrt(pspt*plnu)` to agree with our documentation (Equation (1.2) in the Overleaf "Halo model reaction" section), instead of putting the `sqrt` later.

nebblu replied:
Agree. Have changed it.

Alkistis commented on reactions/src/HALO.cpp:
```diff
@@ -671,6 +715,147 @@ double HALO::plinear_cosmosis(double k) const {
+  psptp = ploopp(vars[0]) + p1hp ;
+
+// cbv terms
+  psptcbv = pspt*plnu;
```
again the sqrt comment here

nebblu replied:
Changed it.

Alkistis commented on reactions/src/HALO.cpp:
```diff
@@ -671,6 +715,147 @@ double HALO::plinear_cosmosis(double k) const {
+    bigE = 1.;
+  }
+  else{
+  kstar = (-0.06/log(argument));
```
would be good to reference Equation (6) in Bose et al here (and in general it will improve user friendliness if important equations are referenced with arXiv number included) -- again, would be much better if instead of 0.06 we had k_bar

nebblu replied:
Added reference.

Alkistis commented on reactions/src/HALO.cpp:
```diff
@@ -671,6 +715,147 @@ double HALO::plinear_cosmosis(double k) const {
+  argument = prefac * (term1 + term2); // 1st root
+// argument = prefac * (term1 - term2); // 2nd root
+
+  if(argument<0.001 || argument>1.){
```
can you comment on what this case corresponds to?

nebblu replied:
Added a comment. This case is to prevent log(-ve number) since for tiny modifications sometimes we end up with a negative argument.

Alkistis commented on reactions/src/SPT.cpp:
```diff
@@ -652,8 +651,29 @@ static double ploopn2_mgdd( const PowerSpectrum& P_L, double vars[], double k, d
+        kv[0] = k;
+        kv[1] = k*r;
+        kv[2] = kv[1];
+        xv[0] = -0.99999999; // don't set to -1 otherwise solver encounters singularity
```
Maybe write this as -1.0 + tol, and define tol as a tolerance parameter for the solver? (also appears later on, so if we change it we need to change it everywhere)

nebblu replied:
Changed all instances in SPT.cpp and added a comment

Alkistis commented on reactions/src/HALO.cpp:
```diff
@@ -247,17 +244,16 @@ static real sigma8d_integrandp_mgcamb(const PowerSpectrum& P, real R, real k) {
 double sig1,sig2;
-  if (vars[6] < 0.5) {
+  if (!mgcamb) {
   // calculate sigma_8^2 and it's smoothing scale derivative @ R=8.
```
useful to make clear what units you're working on, e.g. all k should be h/Mpc thus R should be Mpc/h, all input P(k) must be Mpc^3/h^3, etc.

nebblu replied:
Agree. Since the same units are used throughout, we can probably add this to the readme file.

Alkistis commented on reactions/src/HALO.cpp:
```diff
@@ -355,8 +349,9 @@ return result;
+      double omegacb = vars[1]-vars[6];
       double Deltavir = (1.+ delta_avir(lgmass))*pow3(vars[0]/a_vir(lgmass));
-      return 0.1*pow((pow(10.,lgmass)*Gnewton)/(5.*vars[1]*Deltavir), 1./3.);
+      return 0.1*pow((pow(10.,lgmass)*Gnewton)/(5.*omegacb*Deltavir), 1./3.);
```
can you reference an equation here (what is 0.1?)

nebblu replied:
Have added a comment. The 0.1 pops out upon converting rho_m to the correct units.

Alkistis commented on reactions/src/HALO.cpp:
```diff
@@ -646,6 +641,7 @@ void HALO::react_init(double vars[]) const{
   // real linear spectrum
   plreal = pow2(linear_growth(0.06))*P_l(0.06);
```
pesky 0.06 again here :-D

nebblu replied:
changed throughout :)

Alkistis commented:
Comment on commit 11159ff in #10: creating tests for validating functions is very good but then they are deleted. Would be much better if these were turned into a collection of short and fast unit tests that come together with the code and have to be passed every time the code runs (see e.g. [here](https://docs.python-guide.org/writing/tests/)). Worth thinking about implementing those for future releases. It is extremely useful for bug catching and avoiding, and makes "breaking" the code very hard :-)

Alkistis commented on reactions/src/SPT.cpp:
```diff
@@ -749,6 +789,32 @@ switch (a) {
+    loop = prefac*Integrate<2>(bind(ploopn2_mgdd_nu,cref(P_L),vars,k,std::placeholders::_1,std::placeholders::_2), c, d, err);
+    tree = P_L(k);
+    return loop+tree;
+  case 8:
```
we should document the cases a bit (comment on what they represent -- also, why case 8 after case 1?)

billwright93 replied:
The cases in PLOOPn2_nu() correspond to their equivalents in PLOOPn2(). There is some documentation about what each case represents above PLOOPn2() -- cases 4/5/6/7 correspond to P(k) in IDE but have been removed so the documentation needs updating.

nebblu replied:
Added in comments and changed case 8 to the next integer case consistently in SPT.cpp and HALO.cpp

Alkistis commented on reactions/src/SPT.cpp:
```diff
@@ -712,6 +732,26 @@ static double ploopn2_mgdd_pseudo( const PowerSpectrum& P_L, double vars[], doub
+        p22 = pow2(F2_nk);
+        p13 = F1_nk*F3_nk;
+
+    return pow2(r)*2.*(P_L(k*r)/pow2(F1p_nk))*( (P_L(kargs[0])/pow2(F1kmp_nk))*p22 + 3.*(P_L(k)/pow2(F1_nk))*p13 );
```
@billwright93 where does the "3" come from in the `3.*(P_L(k)/pow2(F1_nk))*p13` term? (I'm probably missing something trivial re the relation to the Overleaf notes)

billwright93 replied:
I've forgotten where this comes from -- will check and get back to you.

billwright93 replied:
So at a glance the code expression doesn't match Eq.(2.25) from Ben's paper (1606.02520) which I've based my notes on. Hopefully @nebblu can explain what's happening here.

nebblu replied:
There is a typo in Eq.2.25 of 1606.02520. In fact, it is inconsistent with 2.20 of 1606.02520 (which is correct).

billwright93 replied:
Ok, I've updated the Overleaf notes to correct the typo.

Alkistis commented on reactions/src/SPT.cpp:
```diff
@@ -843,6 +908,89 @@ for(int zi = 0; zi<noz; zi++){
+  IOW iow;
+
+  double res[noz],resp[noz]; // stores loop integral results for real and pseudo spectra
+  const int loop_N = 50; // steps in trap rule for loop integral in integrated wave vector p
```
have we checked if these steps are enough for the precision needed?

nebblu replied:
I've checked that it gives sub percent accuracy but it could be chosen to be less. I've added a comment.

Alkistis commented on reactions/src/SPT.cpp:
```diff
@@ -652,8 +651,29 @@ static double ploopn2_mgdd( const PowerSpectrum& P_L, double vars[], double k, d
+        xv[0] = -0.99999999; // don't set to -1 otherwise solver encounters singularity
+        xv[1] = x;
+        xv[2] = -x;
+        kargs[0] = sqrt(kv[1]*kv[1]+kv[0]*kv[0]-2.*kv[1]*kv[0]*xv[1]);
```
maybe put a comment here re what this actually is following the Overleaf Equation 2.5 (especially useful if these expressions go into the paper)

billwright93 replied:
So a comment that kargs[0] corresponds to k*sqrt(1+r^2-2rx) in the Overleaf notation?

Alkistis replied:
something like that, yes -- but it's not necessary or urgent, I'm just commenting as I go :-) we can chat about these suggestions with Ben too (I certainly don't have a complete overview of the code, yet)

nebblu replied:
I'll leave this one for the time being. For now have added comments in many places indicating what our integrated k-vector is whicch hopefully makes such expressions more understandable.

nebblu committed:
Made all changes as described in pull comment replies

nebblu committed:
Merge branch 'neutrinos_ben' of https://github.com/nebblu/ReACT into neutrinos_ben

nebblu committed:
Corrected typo in reaction_nu, changed parameters in mass functions to match Matteos for validation

nebblu committed:
Changed initnorm matter density to cb to correctly get sigma_cb,Lambda (eq.1.8 of draft). Need to make sure if mgcamb flag is off, omega_nu=0 ---need to include error flag if not

nebblu committed:
added a validation test code and a folder with MGCAMB output and Matteo's data files

nebblu committed:
Added new A_s dependent validation code and data files produced using Matteo's Mathematica notebook

nebblu committed:
Cleaned initnorm function

nebblu committed:
Added linear power spectrum initialisation that takes A_s as input rather than sigma_8. Additional z-dependent normalisation is provided tthrough the transfer function supplied

nebblu committed:
Commented in the correct root in react_init_nu/2 functions. Also corrected typo in equation for term2 in this function

nebblu committed:
Added Initialise_as function to initialise the normalisation from A_s and added cosmology class which includes a_s as an input parameter

nebblu committed:
Removed the acol dependence from T1, growth d and ainit. Further divide initial boundaries in SphericalCollapse (delta_uu andd delta_ll) by acol - see mysccol , mydelta. This makes the solver completely consistent with Matteo's notebook and allows <0.2% agreement for collapse densities at z=0,1,1.5,3

nebblu commented on reactions/examples/validate.cpp:
```diff
@@ -0,0 +1,183 @@
+#include <cerrno>
```
Validation file with Matteo's output in 'examples/validate' folder

nebblu commented on reactions/src/Cosmology.cpp:
```diff
@@ -26,12 +26,32 @@ Cosmology::Cosmology(real h, real n, real Omega_m, real Omega_b, real sigma8, co
     NormalizeTransferFunction(sigma8);
 }
 
+Cosmology::Cosmology(real h, real n, real Omega_m, real Omega_b, real As, real k0, const array& ki, const array& Ti) {
```
New Cosmology class for As

nebblu commented on reactions/src/Cosmology.cpp:
```diff
@@ -26,12 +26,32 @@ Cosmology::Cosmology(real h, real n, real Omega_m, real Omega_b, real sigma8, co
     Initialize(cosmo);
 }
 
+void Cosmology::Initialize_as(real h_, real n_, real Omega_m_, real Omega_b_, real As_, real k0_) {
```
Initialises A_s + cosmological params

nebblu commented on reactions/src/HALO.cpp:
```diff
@@ -487,9 +497,9 @@ double HALO::cvirial(double lgmass, double acol) const {
-       double q = 0.75;
-       double p = 0.3;
-       double aconst = 1./3.056;
+       double q = 0.7665;//0.75;
```
Matteo's values for validation

nebblu commented on reactions/src/HALO.cpp:
```diff
@@ -57,8 +57,8 @@ void react_error_halo(const char* msg) {
 
 /*Load the linear power spectra */
 // For massless neutrinos or no neutrinos, user should set P_cb = P_nu = P_l (total matter spectrum)
-HALO::HALO(const Cosmology& C_, const PowerSpectrum& P_l_, const PowerSpectrum& P_cb_, const PowerSpectrum& P_nu_, real epsrel_)
```
Require 4 input power spectra - P_m, P_cb, P_nu and P_cb in LCDM

nebblu commented on reactions/src/HALO.h:
```diff
@@ -6,14 +6,16 @@
-// The current values were chosen to satisfy these properties for a wide range of test cases in f(R) gravity where nu is mass dependent. 
+// The current values were chosen to satisfy these properties for a wide range of test cases in f(R) gravity where nu is mass dependent.
 // Ideally you want as small a range as possible so that you are able to solve spherical collapse less often (mass_loop parameter in cosmosis pipeline)
-const double Mmax = 18.;
```
To match Matteo's range of masses

nebblu commented on reactions/src/HALO.cpp:
```diff
@@ -907,10 +923,13 @@ double HALO::reaction_nu(double k, double vars[]) const {
   double php = plm + one_halop(k,vars); // pseudo
 
   // real linear spectrum
- return (fvt2*ph +  fv*fvt*sqrt(ph*plnu) + fv2*plnu)/php;
```
Typo corrected

nebblu commented on reactions/src/HALO.cpp:
```diff
@@ -867,13 +881,13 @@ void HALO::react_init_nu2(double vars[], Spline ploopr, Spline ploopp, bool mgca
   double term1 = p1h*fvt2 + (fvt2*bigE*plcb - fv2*plnu - (p1hp+plm)*rsptk0);
-  double term2 = 2.*sqrt(fvt2*fv2*(p1hp+plm)*plnu*rsptk0);
+  //double term2 = 2.*sqrt(fvt2*fv2*(p1hp+plm)*plnu*rsptk0); // suspected typo -- fvt2 inclusion
+  double term2 = 2.*sqrt(fv2*(p1hp+plm)*plnu*rsptk0);
```
Typo corrected

nebblu commented on reactions/src/HALO.cpp:
```diff
@@ -784,16 +797,17 @@ else{
 
    argument = prefac * (term1 + term2); // 1st root
-// argument = prefac * (term1 - term2); // 2nd root
+  //   argument = prefac * (term1 - term2); // 2nd root
```
1st root is correct

Alkistis replied:
should you add this comment on the actual code (or is it only for a particular case / testing purposes?)

nebblu commented on reactions/src/LinearPS.cpp:
```diff
@@ -80,3 +82,41 @@ real myLinearPS::Evaluate(real k) const {
     else
         return pk(k);
 }
+
```
New linear power spectrum initialisation using As and not sigma8

nebblu commented on reactions/src/SCOL.cpp:
```diff
@@ -60,23 +60,20 @@ static double Pzeta(double x, void * p)
 static int f_modscol(realtype t, N_Vector y, N_Vector ydot, void *user_data)
 {
-    realtype y1, y2, c0, ET, ET0, yenv, Fvir, hubble2, dhlnaoh, prefac, term1, term2, term4, IC, Rth, OM, T1, maxt;
+    realtype y1, y2, ET, ET0, yenv, Fvir, hubble2, dhlnaoh, prefac, prefac2, term1, term2, term4, IC, Rth, OM, OCB, T1, maxt;
```
c0 is unused

nebblu commented on reactions/src/SCOL.cpp:
```diff
@@ -60,23 +60,20 @@ static double Pzeta(double x, void * p)
     data    = (UserData) user_data;
     IC      = data->IC;
     OM      = data->OM;
+    OCB     = data->OCB;
```
addded in new Omega_cb parameter

nebblu commented on reactions/src/SCOL.cpp:
```diff
@@ -264,9 +262,9 @@ double SCOL::Delta_Lambda(double OM, double Z)
   gsl_integration_workspace_free (w);
 
-  return 2.5 * OM_Z_params.a
+  return 2.5 * OM
```
Removed scale factor scaling in growth

nebblu commented on reactions/src/SCOL.cpp:
```diff
@@ -442,8 +441,8 @@ int SCOL::SphericalCollapse(double *dC, arrays_T3 xxyyzz, UserData data_vec, dou
 
-    delta_uu = delta_g;
-    delta_ll = DELTA2;
+    delta_uu =  delta_g;
```
upper and lower limits as in Matteo's notebook

nebblu commented on reactions/src/SCOL.cpp:
```diff
@@ -476,13 +475,14 @@ int SCOL::SphericalCollapse(double *dC, arrays_T3 xxyyzz, UserData data_vec, dou
 
 
-      cvode_mem     = CVodeCreate(CV_ADAMS);
+      cvode_mem     = CVodeCreate(CV_BDF);
```
changed solver to better address stiff ODEs - should probably be changed back to CV_ADAMS...

nebblu commented on reactions/src/SCOL.cpp:
```diff
@@ -613,42 +622,46 @@ double SCOL::myscol(double myscolparams[], double acol, double omega0, double Rt
        gsl_spline *myspline_yenv    = gsl_spline_alloc (gsl_interp_cspline, (*xxyy).count);
        gsl_spline_init (myspline_yenv, (*xxyy).xx, (*xxyy).yy, (*xxyy).count);
 
+      // we progressively solve for final times from af = 0.000045 (Exp[-10]) to acolapse in steps of TMULT, over NOUT_DC steps.
```
made TMULT dependent on number of output steps NOUT_DC

nebblu commented on reactions/src/SCOL.cpp:
```diff
@@ -613,42 +622,46 @@ double SCOL::myscol(double myscolparams[], double acol, double omega0, double Rt
          data->spline  = myspline_yenv;
          data->acc     = acc;
-         data->T1      = T00 + XF;
+         data->T1      = T00; // = Log[ai]
```
Removed scale factor scaling from initial time

nebblu commented on reactions/src/SCOL.cpp:
```diff
@@ -613,42 +622,46 @@ double SCOL::myscol(double myscolparams[], double acol, double omega0, double Rt
         SphericalCollapse (&myscolparams[0], xxyyzz, data, TMULT, mydelta);
        	free(data);
 
-     double ainit = exp(T00)*acol;
```
Removedd scale factor scaling from initial time

nebblu commented on reactions/src/SCOL.cpp:
```diff
@@ -667,33 +680,27 @@ double SCOL::myscol(double myscolparams[], double acol, double omega0, double Rt
         // edit WEFF and mymgF in SpecialFunctions.cpp for different gravity/evolutions - default is LCDM
        double wphi, weff, ke;
-       if(mymg == 1){
+       if(!mymg){
```
changed mymg to a bool operator

nebblu commented on reactions/src/SpecialFunctions.cpp:
```diff
@@ -1088,6 +1088,7 @@ struct param_type2 {
 	real par2;
 	real par3;
   int par4;
+	real par5;
```
added new parameter for omega_nu

nebblu commented on reactions/src/SpecialFunctions.cpp:
```diff
@@ -1101,16 +1102,18 @@ int funcn2(double a, const double G[], double F[], void *params)
+	double hub2 = HA1g(a,omega0,p1,p2,p3);
+	double ap5 = pow(a,5);
 
+	F[0] = G[1];
```
form of new ODE is equivalent to old form but changed for testing sensitivity - also needed to change initial conditions

nebblu commented on reactions/src/SpecialFunctions.cpp:
```diff
@@ -1123,22 +1126,23 @@ double g_de;
-			double a = 0.0001;
-	  	double G1[2] = {a,-a}; // initial conditions
-			struct param_type2 params_my;
+			double a = 3e-5;
```
changed initial conditions because ODE form changed

nebblu committed:
Changed y_environment's acol scaling. Now validated with Matteo's F5 reaction at z=1 and 0

nebblu committed:
changed the halo mass function parameters back to their standard values

nebblu committed:
Updated validate.cpp example to include GR input transfer neededd for pseudo and reaction calculations

nebblu committed:
Added in dustgrain transfers for no massive neutrinos and a output example file

nebblu committed:
Addded in Matteo's transfer functions for validation with arxiv:1812.05594

nebblu committed:
HALO: corrected psptp to eq.2.3 of draft. SPT: made source of linear growth to be Omega_cb for PLOOPn2_nu. SpecialFunctions: Addedd in omega_nu dependence to linear part of evolution equations in initn2 and initn_pseudo.

Alkistis commented on reactions/examples/validate.cpp:
```diff
@@ -49,18 +50,20 @@ string line;
-                  Tnu[i] = mytrans[i][5]/mytrans[0][5]; // should be adjusted if this column is all 0s ....
+                  Tm[i] = mytrans[i][6];
+                  Tcb[i] = mytrans[i][7];
+                  Tnu[i] = mytrans[i][5]; // should be adjusted if this column is all 0s ....
```
I don't understand what this comment means exactly?

nebblu replied:
I was getting some run errors when there was no neeutrino powerr in the transfer functtion. This seems to have gone away for some reason though. Can remove comment.

Alkistis commented on reactions/examples/validate.cpp:
```diff
@@ -156,21 +148,32 @@ string line2;
-//if(fabs(p1-1.)>0.001){
+         p1 = halo.reaction_nu(k,vars)/mypk[i][1];
+
+       // p1 = P_cb(k)/mypk[i][1];
+       // p1 = P_nu(k)/mypk[i][1];
+
+      // p1 = halo.one_halo(k,vars) / mypk[i][1];
+      // p1 = halo.one_halop(k,vars)/ mypk[i][1];
+
+      // p1 = halo.cvirialp(log10(k),vars[0])/ mypk[i][1];
+      // p1 = halo.nvirialp(log10(k),vars[1]-vars[6])/mypk[i][1];
+      // p1 = halo.rvirialp(log10(k),vars)/ mypk[i][1];
+
+//if(fabs(p1-1.)>0.002){
```
are these needed?

nebblu replied:
The criteria of 0.2% agreement or the various quantities? I guess we can keep theem there to make validation of alll quanttitties easier?

Alkistis commented on reactions/examples/halo_ps.cpp:
```diff
@@ -31,17 +36,18 @@ int main(int argc, char* argv[]) {
+double vars[7];
 
-    vars[0] = 1.;
+    vars[0] = 1./(1.+1.);
```
that could be potentially buggy, no? can't you write it in terms of some redshift variable?

nebblu replied:
Yea I agree. Have edited it.

nebblu committed:
made parameter choices clearer

nebblu committed:
improved the param input of spt.cpp example

nebblu committed:
Create EXPERIMENTAL

nebblu committed:
Delete EXPERIMENTAL

nebblu committed:
Update README.md

Commit:
added all relevant functions and introduced a new flag for modified gravity or not to set kstar and \mathcal{E} to their LCDM values if it is false

Commit:
removed dustgrain transfers and added an example file with wcdm and massive neutrinos

Commit:
corrected transfer path

Commit:
edited halo_ps.cpp with new flags

Commit:
removed matteo's data files

Commit:
added default values for flags so that pyreact doesn't break. Added benchmarks in tests for wcdm+mnu and fr+mnu

Commit:
updated README

nebblu commented:
Hi Bill and Alkistis, 

Have added a few features and some tests. The main new updates are:

Updated the reac_init_nu to remove a derivation mistake. 

Updated relevant functions with a new flag - modg - that signals modified gravity being active or not (this is needed to set kstar and \mathcal{E} to LCDM values forcefully). 

Updated header files to include default flag values (mgcamb and modg) so that the old python wrapper doesn't break. This should be updated at some point. 

Cleaned the example folder to include a reduced example set. 

Included some tests in tests/tests_mnu which will check if the output you are getting is consistent with what I get now (compares against the wcdm and f(R) + m_nu output I got for our paper). 

If everything looks ok, we get merge and then have some people test it. I should also update the python wrapper and cosmosis interface at some point.

Commit:
updated halo class to include 4 pl inputs in  react_wrapper.cpp

Commit:
added a fix to DELTA1 in spherical collapse that allows us to solve for high fr0 and high neutrino mass

billwright93 commented on reactions/src/HALO.cpp:
```diff
@@ -482,9 +476,14 @@ double HALO::cvirial(double lgmass, double acol) const {
+       // Quantities as in https://arxiv.org/abs/1909.02561
+       // double q = 0.7665;
+       // double p = 0.2488;
+       // double aconst = 0.3292;
```
Why have these parameters changed and where do the new values come from?

nebblu replied:
The first set are the standard N-body fits used with the ST mass function. The second set are those fit to the simulations used in https://arxiv.org/abs/1909.02561 which I used to validate the neutrino code with Matteo's data.

billwright93 replied:
Ah ok, and I see now above this it refers to Eq. (40) in 1812.05594 which is great. Is there another paper that we can cite there that explains how these standard values were identified e.g. the original ST papers?

nebblu replied:
Yea definitely, will add that in!

billwright93 commented on reactions/src/HALO.cpp:
```diff
@@ -500,9 +499,14 @@ double HALO::cvirial(double lgmass, double acol) const {
+       // Quantities as in https://arxiv.org/abs/1909.02561
+       // double q = 0.7665;
+       // double p = 0.2488;
+       // double aconst = 0.3292;
```
Same as above.

billwright93 commented on reactions/src/HALO.cpp:
```diff
@@ -730,6 +746,10 @@ void HALO::react_init2(double vars[], Spline ploopr, Spline ploopp) const{
 
 // vars is 0: scale factor, 1: omega_matter today , 2: modified gravity param
 double HALO::reaction(double k, double vars[]) const {
+  // Set to 1 for very large scales
+  if (k<1e-3) {
+    return 1.;
+  }
```
Remind me why we force the reaction to 1 for very large scales - is it simply for speed or does the actual reaction computation not give 1 for large  scales?

nebblu replied:
The reaction doesn't go to exactly 1 at large scales for some numerical reasons (up to 0.5% difference)

billwright93 replied:
Ok, thanks.

Alkistis replied:
Would add this explicitly as a comment (we force the reaction to 1 for very large scales due to numerical reasons that can result to a 0.5% diff).

billwright93 commented on reactions/src/HALO.cpp:
```diff
@@ -913,15 +949,70 @@ void HALO::react_init_nu2(double vars[], Spline ploopr, Spline ploopp, bool mgca
-// Calculate the reaction with massive neutrinos
+/* The 1-loop SPT reaction */
+
+double HALO::reaction_spt(double k0, double vars[], bool mgcamb) const{
```
What is this used for? I can't find an example.

nebblu replied:
We use SPT at 0.06h/Mpc so it should agree exactly with the halo model reaction there. I used this function to test that for the MG+neutrino cases. It's not needed really, but might be useful to keep.

billwright93 replied:
Ok, thanks.

billwright93 commented on reactions/src/HALO.cpp:
```diff
@@ -913,15 +949,70 @@ void HALO::react_init_nu2(double vars[], Spline ploopr, Spline ploopp, bool mgca
+
+// Calculate the reaction with massive neutrinos
 double HALO::reaction_nu(double k, double vars[]) const {
+  if (k<1e-3) {
+    return 1.;
+  }
```
Same as above.

billwright93 commented on reactions/src/SPT.cpp:
```diff
@@ -781,22 +782,20 @@ static double ploopn2_mgdd_nu( const PowerSpectrum& P_L, double vars[], double k
         kargs[1] = sqrt(kv[2]*kv[2]+2.*kv[2]*kv[0]*xv[2]+kv[0]*kv[0]);
         kargs[3] = sqrt(kv[0]*kv[0]+2.*kv[0]*kv[1]*xv[1]+kv[1]*kv[1]);
-        iow.initn2(vars[0],kv,xv,kargs,omegacb,vars[2],vars[3],vars[4],vars[6]);
+        iow.initn2(vars[0],kv,xv,kargs,vars[1],vars[2],vars[3],vars[4],vars[6]);
```
Can you remind me why we now use Omega_m instead of Omega_cb here? I can't see any change to initn2 that matches this, although I see vars[6] is now Omega_cb.

nebblu replied:
So we calculate omega_cb in funcn1 and funcn1_pseudo in SpecialFunctions.cpp using omega0 and omeganu.

billwright93 replied:
Ah right, for some reason I thought vars[6] was Omega_cb but as you say it is Omega_nu - my bad!

billwright93 commented on reactions/src/SPT.cpp:
```diff
@@ -936,7 +937,7 @@ for(int zi = 0; zi<noz; zi++){
 }
 
-// BILL MOD
+//16/03/2021:  Work in progress - needs to be updated to be used in cosmosis
```
Is this currently broken? What happens if a user tries to call this via cosmosis -- perhaps it should throw an error or warning if it doesn't already?

nebblu replied:
So Cosmosis shouldn't break but you can only use it for MG only. It won't break since I've allowed the necessary cosmosis wrapper functions to assume a LCDM input P(k) (whereas for massive neutrinos you need the true transfer function to be fed). 

Implementation massive neutrinos to cosmosis will be more involved since we need to feed in the transfer functions for total matter, cb and massive neutrinos for all redshifts which will require us editing the camb module in cosmosis.

billwright93 replied:
Ok, thanks.

billwright93 commented on reactions/src/SpecialFunctions.cpp:
```diff
@@ -451,11 +451,6 @@ int funcn1(double a, const double G[], double F[], void *params)
 	mu2 = mu(a,k2,omega0,p1,p2,p3);
 	mu3 = mu(a,karg1,omega0,p1,p2,p3);
 
```
Why no mu4=mu(a,karg4,omega0,p1,p2,p3); to use in F[9] as in funcn1_psuedo?

nebblu replied:
It was added explicitly into F[9] - have changed it to be consistent with funcn1_pseudo.

billwright93 commented on reactions/src/SpecialFunctions.cpp:
```diff
@@ -561,10 +556,10 @@ int funcn1_pseudo(double a, const double G[], double F[], void *params)
-	mu2 = mu(a,k2,omega0,p1,p2,p3);
-	mu3 = mu(a,karg1,omega0,p1,p2,p3);
-
+	mu1 = 1;//mu(a,k1,omega0,p1,p2,p3);
+	mu2 = 1;//mu(a,k2,omega0,p1,p2,p3);
+	mu3 = 1;//mu(a,karg1,omega0,p1,p2,p3);
+	mu4 = 1;//mu(a,karg4,omega0,p1,p2,p3);
```
These are hardcoded to 1 rather than using mu because it's the pseudo case right? Is there ever a situation where you'd want to uncomment the mu() here?

nebblu replied:
Yea you might want to use mu, for example in the unscreened approximation that Matteo used in his first reaction paper. I'll re arrange this and include a comment.

billwright93 reviewed (commented):
I have a few questions, but otherwise it looks good!

billwright93 commented on reactions/src/SPT.cpp:
```diff
@@ -633,58 +632,68 @@ real SPT::Lag_bias(int a, real k, real bias[]) const {
         kargs[1] = sqrt(kv[2]*kv[2]+2.*kv[2]*kv[0]*xv[2]+kv[0]*kv[0]);
         kargs[3] = sqrt(kv[0]*kv[0]+2.*kv[0]*kv[1]*xv[1]+kv[1]*kv[1]);
-        iow.initn2(vars[0],kv,xv,kargs,vars[1],vars[2],vars[3],vars[4]);
+        iow.initn2(vars[0],kv,xv,kargs,vars[1],vars[2],vars[3],vars[4],0.);
```
Why do we use Omega_nu=0 here instead of vars[6] as in ploopn2_mgdd?

nebblu replied:
We don't use the velocity spectra currently so left total matter as a source, but agree for consistency we should change that.

billwright93 commented on reactions/src/SPT.cpp:
```diff
@@ -633,58 +632,68 @@ real SPT::Lag_bias(int a, real k, real bias[]) const {
         kargs[1] = sqrt(kv[2]*kv[2]+2.*kv[2]*kv[0]*xv[2]+kv[0]*kv[0]);
         kargs[3] = sqrt(kv[0]*kv[0]+2.*kv[0]*kv[1]*xv[1]+kv[1]*kv[1]);
-        iow.initn2(vars[0],kv,xv,kargs,vars[1],vars[2],vars[3],vars[4]);
+        iow.initn2(vars[0],kv,xv,kargs,vars[1],vars[2],vars[3],vars[4],0.);
```
As above, why do we use Omega_nu=0 here instead of vars[6] as in ploopn2_mgdd?

nebblu replied:
see above

Commit:
added comment on halo mass function default values

Commit:
changed 1-loop velocity spectra to be sourced by cb instead of total matter

Commit:
added unscreened approximation option in funcn1_pseudo

Commit:
added mu4 to funcn1

Commit:
removed multiple z initialisation functions until we've figured out how to use them with cosmosis and a single camb call

billwright93 commented on reactions/src/HALO.cpp:
```diff
@@ -1188,7 +1194,7 @@ double HALO::PHALO_pseudo(double k, bool mgcamb) const{
   }
 
-  if (phpars_pseudo[0] == 1000. || k<=0.005) {
+  if (phpars_pseudo[0] == 1000. || k<=0.01) {
```
Sorry I'm really nitpicking now, but what does this line do? And should 0.01 be a named variable e.g. k_max or something a bit more descriptive?

nebblu replied:
Ah this is some selection criteria to transition from completely linear to non-linear for halofit. Details would be in Takashi et al 1208.2701. I don't think these need to be commented on so much, I have included the Takashi et al reference at the top of the formula for details.

billwright93 replied:
Ah ok, thanks.

Commit:
removed test edit from halofit formula

Commit:
set sigma8 integral top limit back to 50.

Commit:
added more comments on mgcamb and modg flags

Commit:
added comments on the choices for initial delta and cases where a non-standard choice  is applicable

Commit:
added a consistency check against cases with  Omega_nu != 0 with mgcamb = false

Commit:
Updated README

billwright93 reviewed (approved):
This branch looks in great shape and there are lots of comments explaining the changes and how to use the new functionality.

nebblu commented:
The neutrinos branch can be deleted and we can work from neutrinos_ben from now.
